File: src/utils/emailParser.js
// src/utils/emailParser.js
import PostalMime from "postal-mime";

const URL_REGEX = /https?:\/\/[^\s<>"'\)}\]>]+/gi;
const EMAIL_REGEX = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/gi;
const IP_REGEX =
  /\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b/g;
const DOMAIN_REGEX =
  /(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}/gi;

function generateId(content) {
  const s = typeof content === "string" ? content : "";
  let hash = 0;
  for (let i = 0; i < s.length; i++) {
    const char = s.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash;
  }
  return Math.abs(hash).toString(16).substring(0, 16);
}

function extractHeader(headers, name) {
  const header = headers.find((h) => h.key.toLowerCase() === name.toLowerCase());
  return header ? header.value : null;
}

function parseAuthenticationResults(headers) {
  const authResults = extractHeader(headers, "authentication-results") || "";
  const receivedSpf = extractHeader(headers, "received-spf") || "";
  const dkimSignature = extractHeader(headers, "dkim-signature");

  const spfStatus = (() => {
    const combined = (authResults + " " + receivedSpf).toLowerCase();
    if (
      combined.includes("spf=pass") ||
      (receivedSpf.toLowerCase().includes("pass") && !receivedSpf.toLowerCase().includes("fail"))
    )
      return "pass";
    if (combined.includes("spf=fail") || receivedSpf.toLowerCase().includes("fail")) return "fail";
    if (combined.includes("spf=softfail") || receivedSpf.toLowerCase().includes("softfail"))
      return "softfail";
    if (combined.includes("spf=neutral") || receivedSpf.toLowerCase().includes("neutral"))
      return "neutral";
    return "unknown";
  })();

  const dkimStatus = (() => {
    const lower = authResults.toLowerCase();
    if (lower.includes("dkim=pass")) return "pass";
    if (lower.includes("dkim=fail")) return "fail";
    if (dkimSignature) return "present";
    return "unknown";
  })();

  const dmarcStatus = (() => {
    const lower = authResults.toLowerCase();
    if (lower.includes("dmarc=pass")) return "pass";
    if (lower.includes("dmarc=fail")) return "fail";
    if (lower.includes("dmarc=none")) return "none";
    return "unknown";
  })();

  return {
    spf: receivedSpf || null,
    dkim: dkimSignature || null,
    dmarc: authResults || null,
    spf_status: spfStatus,
    dkim_status: dkimStatus,
    dmarc_status: dmarcStatus,
  };
}

function extractUrls(text, html) {
  const urls = new Set();

  (text.match(URL_REGEX) || []).forEach((url) => urls.add(url.replace(/[.,)\]>;]+$/, "")));
  (html.match(URL_REGEX) || []).forEach((url) => urls.add(url.replace(/[.,)\]>;]+$/, "")));

  const hrefRegex = /href=["']([^"']+)["']/gi;
  let match;
  while ((match = hrefRegex.exec(html)) !== null) {
    if (match[1].startsWith("http")) urls.add(match[1]);
  }

  return Array.from(urls);
}

function extractDomains(urls, text) {
  const domains = new Set();

  urls.forEach((urlStr) => {
    try {
      const url = new URL(urlStr);
      domains.add(url.hostname.toLowerCase());
    } catch {
      // ignore
    }
  });

  (text.match(DOMAIN_REGEX) || []).forEach((domain) => {
    const lower = domain.toLowerCase();
    if (
      !lower.endsWith(".png") &&
      !lower.endsWith(".jpg") &&
      !lower.endsWith(".gif") &&
      !lower.endsWith(".css")
    ) {
      domains.add(lower);
    }
  });

  return Array.from(domains);
}

function extractIps(headers) {
  const ips = new Set();

  headers.forEach((header) => {
    (header.value.match(IP_REGEX) || []).forEach((ip) => {
      if (
        !ip.startsWith("10.") &&
        !ip.startsWith("192.168.") &&
        !ip.startsWith("127.") &&
        !ip.startsWith("0.")
      ) {
        ips.add(ip);
      }
    });
  });

  return Array.from(ips);
}

function extractEmailAddresses(text, headers) {
  const emails = new Set();

  (text.match(EMAIL_REGEX) || []).forEach((email) => emails.add(email.toLowerCase()));
  headers.forEach((header) => {
    (header.value.match(EMAIL_REGEX) || []).forEach((email) => emails.add(email.toLowerCase()));
  });

  return Array.from(emails);
}

async function hashAttachment(dataUint8) {
  const hashBuffer = await crypto.subtle.digest("SHA-256", dataUint8);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
}

function base64ToUint8Array(base64) {
  const clean = base64.replace(/\s/g, "");
  const binary = atob(clean);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function normalizeToUint8Array(input) {
  if (input instanceof Uint8Array) return input;
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (typeof input === "string") return new TextEncoder().encode(input);
  return new TextEncoder().encode(String(input ?? ""));
}

function normalizeAttachmentContent(att) {
  if (att?.content instanceof Uint8Array) return att.content;
  if (att?.content instanceof ArrayBuffer) return new Uint8Array(att.content);

  // Some parsers provide base64 strings
  if (typeof att?.content === "string") {
    try {
      return base64ToUint8Array(att.content);
    } catch {
      return new TextEncoder().encode(att.content);
    }
  }

  // e.g. { buffer: ArrayBuffer }
  if (att?.content?.buffer instanceof ArrayBuffer) return new Uint8Array(att.content.buffer);

  return new Uint8Array();
}

function isLikelyPdf(contentType, filename) {
  const ct = (contentType || "").toLowerCase();
  const fn = (filename || "").toLowerCase();
  return ct.includes("pdf") || fn.endsWith(".pdf");
}

function validatePdfHeader(bytes) {
  // PDF must start with "%PDF-"
  if (!bytes || bytes.byteLength < 5) return false;
  const head = new TextDecoder().decode(bytes.slice(0, 5));
  return head === "%PDF-";
}

export async function parseEmail(content) {
  const parser = new PostalMime();

  // ✅ Parse bytes for reliable attachment decoding
  const rawBytes = normalizeToUint8Array(content);
  const parsed = await parser.parse(rawBytes);

  const headers = (parsed.headers || []).map((h) => ({
    name: h.key,
    value: h.value,
  }));

  const bodyText = parsed.text || "";
  const bodyHtml = parsed.html || "";

  const attachments = await Promise.all(
    (parsed.attachments || []).map(async (att) => {
      const bytes = normalizeAttachmentContent(att);
      const contentType = att.mimeType || "application/octet-stream";
      const filename = att.filename || "unknown";

      const sha256 = await hashAttachment(bytes);
      const blob = new Blob([bytes], { type: contentType });

      // ✅ Stronger UX: detect “PDF that isn't really a PDF”
      let preview_error = null;
      if (isLikelyPdf(contentType, filename) && !validatePdfHeader(bytes)) {
        preview_error =
          "Attachment is labeled PDF but does not contain a valid PDF header (%PDF-). " +
          "Bytes may be truncated or decoded incorrectly.";
      }

      return {
        filename,
        content_type: contentType,
        size: bytes.byteLength,
        sha256,
        blob,
        preview_error,
      };
    })
  );

  const authentication = parseAuthenticationResults(parsed.headers || []);
  const urls = extractUrls(bodyText, bodyHtml);
  const domains = extractDomains(urls, bodyText);
  const ip_addresses = extractIps(parsed.headers || []);
  const email_addresses = extractEmailAddresses(bodyText, parsed.headers || []);

  return {
    id: generateId(typeof content === "string" ? content : ""),
    subject: parsed.subject || "",
    from: parsed.from?.address
      ? `${parsed.from.name || ""} <${parsed.from.address}>`.trim()
      : "",
    to: parsed.to?.map((t) => t.address).join(", ") || "",
    reply_to: parsed.replyTo?.map((r) => r.address).join(", ") || null,
    return_path: extractHeader(parsed.headers || [], "return-path"),
    date: parsed.date ? new Date(parsed.date).toISOString() : null,
    headers,
    body_text: bodyText,
    body_html: bodyHtml,
    urls,
    domains,
    ip_addresses,
    email_addresses,
    attachments,
    authentication,
    raw_content: content,
  };
}

File: src-tauri/src/email_parser.rs
use mailparse::{parse_mail, MailHeaderMap, ParsedMail};
use regex::Regex;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashSet;
use once_cell::sync::Lazy;

static URL_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"https?://[^\s<>"'\)}\]>]+"#).unwrap()
});

static EMAIL_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"#).unwrap()
});

static IP_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b"#).unwrap()
});

static DOMAIN_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}"#).unwrap()
});

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct EmailHeader {
    pub name: String,
    pub value: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Attachment {
    pub filename: String,
    pub content_type: String,
    pub size: usize,
    pub sha256: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AuthenticationResult {
    pub spf: Option<String>,
    pub dkim: Option<String>,
    pub dmarc: Option<String>,
    pub spf_status: String,
    pub dkim_status: String,
    pub dmarc_status: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ParsedEmail {
    pub id: String,
    pub subject: String,
    pub from: String,
    pub to: String,
    pub reply_to: Option<String>,
    pub return_path: Option<String>,
    pub date: Option<String>,
    pub headers: Vec<EmailHeader>,
    pub body_text: String,
    pub body_html: String,
    pub urls: Vec<String>,
    pub domains: Vec<String>,
    pub ip_addresses: Vec<String>,
    pub email_addresses: Vec<String>,
    pub attachments: Vec<Attachment>,
    pub authentication: AuthenticationResult,
    pub raw_content: String,
}

fn extract_header(mail: &ParsedMail, name: &str) -> Option<String> {
    mail.headers.get_first_value(name)
}

fn parse_authentication_results(mail: &ParsedMail) -> AuthenticationResult {
    let auth_results = extract_header(mail, "Authentication-Results").unwrap_or_default();
    let received_spf = extract_header(mail, "Received-SPF").unwrap_or_default();
    let dkim_signature = extract_header(mail, "DKIM-Signature");

    let spf_status = if auth_results.to_lowercase().contains("spf=pass") || received_spf.to_lowercase().contains("pass") {
        "pass".to_string()
    } else if auth_results.to_lowercase().contains("spf=fail") || received_spf.to_lowercase().contains("fail") {
        "fail".to_string()
    } else if auth_results.to_lowercase().contains("spf=softfail") || received_spf.to_lowercase().contains("softfail") {
        "softfail".to_string()
    } else if auth_results.to_lowercase().contains("spf=neutral") || received_spf.to_lowercase().contains("neutral") {
        "neutral".to_string()
    } else {
        "unknown".to_string()
    };

    let dkim_status = if auth_results.to_lowercase().contains("dkim=pass") {
        "pass".to_string()
    } else if auth_results.to_lowercase().contains("dkim=fail") {
        "fail".to_string()
    } else if dkim_signature.is_some() {
        "present".to_string()
    } else {
        "unknown".to_string()
    };

    let dmarc_status = if auth_results.to_lowercase().contains("dmarc=pass") {
        "pass".to_string()
    } else if auth_results.to_lowercase().contains("dmarc=fail") {
        "fail".to_string()
    } else if auth_results.to_lowercase().contains("dmarc=none") {
        "none".to_string()
    } else {
        "unknown".to_string()
    };

    AuthenticationResult {
        spf: Some(received_spf).filter(|s| !s.is_empty()),
        dkim: dkim_signature,
        dmarc: Some(auth_results.clone()).filter(|s| !s.is_empty()),
        spf_status,
        dkim_status,
        dmarc_status,
    }
}

fn extract_body_parts(mail: &ParsedMail) -> (String, String) {
    let mut text_body = String::new();
    let mut html_body = String::new();

    fn process_part(part: &ParsedMail, text: &mut String, html: &mut String) {
        let content_type = part.ctype.mimetype.to_lowercase();

        if content_type.starts_with("multipart/") {
            for subpart in &part.subparts {
                process_part(subpart, text, html);
            }
        } else if content_type == "text/plain" {
            if let Ok(body) = part.get_body() {
                *text = body;
            }
        } else if content_type == "text/html" {
            if let Ok(body) = part.get_body() {
                *html = body;
            }
        }
    }

    process_part(mail, &mut text_body, &mut html_body);

    if text_body.is_empty() && !html_body.is_empty() {
        text_body = strip_html(&html_body);
    }

    (text_body, html_body)
}

fn strip_html(html: &str) -> String {
    let tag_regex = Regex::new(r"<[^>]+>").unwrap();
    let result = tag_regex.replace_all(html, " ");
    let whitespace_regex = Regex::new(r"\s+").unwrap();
    whitespace_regex.replace_all(&result, " ").trim().to_string()
}

fn extract_attachments(mail: &ParsedMail) -> Vec<Attachment> {
    let mut attachments = Vec::new();

    fn process_part(part: &ParsedMail, attachments: &mut Vec<Attachment>) {
        let content_type = part.ctype.mimetype.to_lowercase();

        if content_type.starts_with("multipart/") {
            for subpart in &part.subparts {
                process_part(subpart, attachments);
            }
        } else {
            let disposition = part.headers.get_first_value("Content-Disposition").unwrap_or_default();
            if disposition.contains("attachment") ||
               (!content_type.starts_with("text/") && !content_type.starts_with("multipart/")) {
                if let Ok(body) = part.get_body_raw() {
                    let filename = part.ctype.params.get("name")
                        .cloned()
                        .or_else(|| {
                            let re = Regex::new(r#"filename="?([^";\s]+)"?"#).unwrap();
                            re.captures(&disposition).map(|c| c[1].to_string())
                        })
                        .unwrap_or_else(|| "unknown".to_string());

                    let mut hasher = Sha256::new();
                    hasher.update(&body);
                    let hash = hex::encode(hasher.finalize());

                    attachments.push(Attachment {
                        filename,
                        content_type: part.ctype.mimetype.clone(),
                        size: body.len(),
                        sha256: hash,
                    });
                }
            }
        }
    }

    process_part(mail, &mut attachments);
    attachments
}

fn extract_urls(text: &str, html: &str) -> Vec<String> {
    let mut urls: HashSet<String> = HashSet::new();

    for cap in URL_REGEX.find_iter(text) {
        urls.insert(cap.as_str().trim_end_matches(&['.', ',', ')', ']', '>', ';'][..]).to_string());
    }

    for cap in URL_REGEX.find_iter(html) {
        urls.insert(cap.as_str().trim_end_matches(&['.', ',', ')', ']', '>', ';'][..]).to_string());
    }

    let href_regex = Regex::new(r#"href=["']([^"']+)["']"#).unwrap();
    for cap in href_regex.captures_iter(html) {
        let url = &cap[1];
        if url.starts_with("http") {
            urls.insert(url.to_string());
        }
    }

    urls.into_iter().collect()
}

fn extract_domains(urls: &[String], text: &str) -> Vec<String> {
    let mut domains: HashSet<String> = HashSet::new();

    for url_str in urls {
        if let Ok(parsed_url) = url::Url::parse(url_str) {
            if let Some(host) = parsed_url.host_str() {
                domains.insert(host.to_lowercase());
            }
        }
    }

    for cap in DOMAIN_REGEX.find_iter(text) {
        let domain = cap.as_str().to_lowercase();
        if !domain.ends_with(".png") && !domain.ends_with(".jpg") &&
           !domain.ends_with(".gif") && !domain.ends_with(".css") {
            domains.insert(domain);
        }
    }

    domains.into_iter().collect()
}

fn extract_ips(headers: &[EmailHeader]) -> Vec<String> {
    let mut ips: HashSet<String> = HashSet::new();

    for header in headers {
        for cap in IP_REGEX.find_iter(&header.value) {
            let ip = cap.as_str();
            if !ip.starts_with("10.") && !ip.starts_with("192.168.") &&
               !ip.starts_with("127.") && !ip.starts_with("0.") {
                ips.insert(ip.to_string());
            }
        }
    }

    ips.into_iter().collect()
}

fn extract_email_addresses(text: &str, headers: &[EmailHeader]) -> Vec<String> {
    let mut emails: HashSet<String> = HashSet::new();

    for cap in EMAIL_REGEX.find_iter(text) {
        emails.insert(cap.as_str().to_lowercase());
    }

    for header in headers {
        for cap in EMAIL_REGEX.find_iter(&header.value) {
            emails.insert(cap.as_str().to_lowercase());
        }
    }

    emails.into_iter().collect()
}

pub fn parse_eml(content: &str) -> Result<ParsedEmail, String> {
    let mail = parse_mail(content.as_bytes())
        .map_err(|e| format!("Failed to parse email: {}", e))?;

    let headers: Vec<EmailHeader> = mail.headers.iter()
        .map(|h| EmailHeader {
            name: h.get_key().to_string(),
            value: h.get_value(),
        })
        .collect();

    let (body_text, body_html) = extract_body_parts(&mail);
    let attachments = extract_attachments(&mail);
    let authentication = parse_authentication_results(&mail);

    let urls = extract_urls(&body_text, &body_html);
    let domains = extract_domains(&urls, &body_text);
    let ip_addresses = extract_ips(&headers);
    let email_addresses = extract_email_addresses(&body_text, &headers);

    let mut hasher = Sha256::new();
    hasher.update(content.as_bytes());
    let id = hex::encode(&hasher.finalize()[..8]);

    Ok(ParsedEmail {
        id,
        subject: extract_header(&mail, "Subject").unwrap_or_default(),
        from: extract_header(&mail, "From").unwrap_or_default(),
        to: extract_header(&mail, "To").unwrap_or_default(),
        reply_to: extract_header(&mail, "Reply-To"),
        return_path: extract_header(&mail, "Return-Path"),
        date: extract_header(&mail, "Date"),
        headers,
        body_text,
        body_html,
        urls,
        domains,
        ip_addresses,
        email_addresses,
        attachments,
        authentication,
        raw_content: content.to_string(),
    })
}
